bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java

```
@@ -567,39 +573,65 @@ private String getDataDir(Display display) {
 	long[] ppv = new long[1];
 	int hr = containingEnvironment.environment().QueryInterface(COM.IID_ICoreWebView2Environment2, ppv);
 	if (hr == COM.S_OK) environment2 = new ICoreWebView2Environment2(ppv[0]);
 	// The webview calls are queued to be executed when it is done executing the current task.
-	IUnknown setupBrowserCallback = newCallback((result, pv) -> {
-		if ((int)result == COM.S_OK) {
+	containingEnvironment.environment().CreateCoreWebView2Controller(browser.handle, createControllerInitializationCallback(previousAttempts));
+}
+
+private IUnknown createControllerInitializationCallback(int previousAttempts) {
+	Runnable initializationRollback = () -> {
+		webViewProvider.abortInitialization();
+		if (environment2 != null) {
+			environment2.Release();
+			environment2 = null;
+		}
+		containingEnvironment.instances().remove(this);
+	};
+	return newCallback((result, pv) -> {
+		if (browser.isDisposed()) {
+			initializationRollback.run();
+			return COM.S_OK;
+		}
+		if (result == OS.HRESULT_FROM_WIN32(OS.ERROR_INVALID_STATE)) {
+			initializationRollback.run();
+			SWT.error(SWT.ERROR_INVALID_ARGUMENT, null,
+					" Edge instance with same data folder but different environment options already exists");
+		}
+		switch ((int) result) {
+		case COM.S_OK:
 			new IUnknown(pv).AddRef();
+			setupBrowser((int) result, pv);
+			break;
+		case COM.E_WRONG_THREAD:
+			initializationRollback.run();
+			error(SWT.ERROR_THREAD_INVALID_ACCESS, (int) result);
+			break;
+		case COM.E_ABORT:
+			initializationRollback.run();
+			break;
+		default:
+			initializationRollback.run();
+			if (previousAttempts < MAXIMUM_CREATION_RETRIES) {
+				System.err.println(String.format("Edge initialization failed, retrying (attempt %d / %d)", previousAttempts + 1, MAXIMUM_CREATION_RETRIES));
+				createInstance(previousAttempts + 1);
+			} else {
+				System.err.println(String.format("Aborting Edge initialiation after %d retries", MAXIMUM_CREATION_RETRIES));
```

Wondering if we should also throw an error here since the initialization was not successful.

---

Yes, makes sense. I adapted this to throw an error instead of just printing an error message with ca41f1e.